New Use Case about CNF behavior in respect to infrastructure lifecycle events (#95)
cncf/cnf-wg: use-case/0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md

# UC-0001: Lifecycle management of infrastructure where CNF runs

> ## Glossary

Reviewer: ```suggestion
## Glossary

```

- CNF - Cloud Native Network Function

> - Cloud Native Platform - Platform that exhibits properties of the cloud on which CNF runs. For this use case Cloud Native Platform = Kubernetes cluste

Reviewer 2: ```suggestion
- Cloud Native Platform - Platform that exhibits properties of the cloud on which CNF runs. For this use case Cloud Native Platform = Kubernetes cluster
```

## Involved processes
- [ ] Development
- [ ] Deployment
- [ ] System integration
- [ ] Network integration

> - [x] Lifecycle management

Reviewer 2: Is this LCM the LCM of the infrastructure or the application or both?

Author: It is infrastructure lifecycle management. It is suggested in the title.

- one of existing nodes gets cordoned, than drained and finally eliminated from the cluster
- this process repeats until all old nodes are eliminated from the cluster and replaced with fresh nodes

This process is first done in test environment, followed by reference and finally production environment with certain time distance (typically 7-10 days). 

Cloud Native Platform DevOps team expects that they can do the changes on infrastructure "in place", while CNF is operating, and in regular working hours (no maintenance windows). They count that CNF DevOps team can detect anomalies on CNF after changes on infrastructure and report them early, ideally in test environment. Finally they expect that if anomalies are detected they can jointly troubleshoot and eliminate them together with CNF DevOps team in order to proceed with infrastructure changes asap. 

CNF DevOps team expects that changes on infrastructure will not cause neither downtime of CNF in production nor significant and lasting degradation of its function or performance. They could however tolerate some temporary degradation of performance during process of change (e.g. some request retries etc)

## Expected behaviour

> It is expected that CNF copes well with situations when one or more nodes in the cluster are replaced with fresh nodes. CNF shall be capable to bear, without visible impact on its function or performance, the situations in which its workloads are shifted to the available nodes. It is expected that propper functioning of CNF is not tight with any single node in the cluster or group of nodes in bigger clusters.

Reviewer 2: Maybe "replaced with fresh nodes" should be extended to "replaced with fresh nodes with similar or better capabilities"

Author: correct!

> It is expected that CNF copes well with situations when one or more nodes in the cluster are replaced with fresh nodes. CNF shall be capable to bear, without visible impact on its function or performance, the situations in which its workloads are shifted to the available nodes. It is expected that propper functioning of CNF is not tight with any single node in the cluster or group of nodes in bigger clusters.

Reviewer 2: CNF functionality can depend on not capabilities, like instruction sets or HW accelerators. In some cases these capabilities are available only in a group of nodes.

Author: true. the main point here is that if CNF pods run on such node e.g. with HW accelerators CNF will not have problems if that node disappears and node with same capabilities appears or if CNF pods are shifted by Kubernetes scheduler to another node with same capabilities.

A CNF should be automatically testable, with regards to its proper functionality, in a relatively short time after changes to the infrastructure. Ideally, in the test and reference environments, there would be nightly tests running. The CNF team would be informed each day what, if any, tests broke their CNF. This means that it shall have automated test coverage at the level of 100%.  

> 

Author: Please check the latest version

> 

Reviewer 3: Paints a clearer picture. Will do a quick editing pass, but the point is more distinct.

> 

Reviewer: @jeffsaelens resolving to focus comments on latest version.  Re-open if you think these conversation thread is relevant to the current version.

> 

Reviewer 4: @jeffsaelens just wanted to bump to see if you are happy with how it is now

> 

Reviewer 3: I think we are good to push this in its current state and continue to evaluate as we dive deeper.

> It is expected that a CNF copes well with situations when one or more nodes in the cluster are replaced with fresh nodes with same or better capabilities. Assuming that remaining capacity of the cluster is allowing rescheduling CNF pods to other, equivalent nodes, the CNF should be capable of bearing, without visible impact on its own function or performance, the situations in which it's workloads are shifted to those nodes. If, within thecourse of infrastructure lifecycle operations, the capacity of the cluster gets reduced but remains above minimally allowed level, CNF shall still function with graceful degradation of its performance in the worst case. It is an expected functionality, that a proper CNF is not tightly coupled to any single node in a cluster, or group of nodes in larger clusters. CNF can still demand that nodes with certain capabilities are present (e.g. SR-IOV, XDP, GPU etc). However it shall not have issues nor impact for the customers if its worlkoads are dynamically shifted by kube-scheduler from one to another node of such type.

Reviewer 5: I think this use case is a perfect example for highlighting the portability as best practice.

## Challenges and limitations with Kube-native approach

> Most of today's (so called) CNFs are retrofitted VNFs that have been forklifted into the containers. They are frequently pinning the nodes. CNFs expect that the cluster nodes are long living, even highly available. In Kube-native approach cluster nodes are actually ephemeral and relatively short living. They exist typically for couple of days to couple of weeks. In such situation CNFs are implicitly requiring that infrastructure is mutable and that Cloud Native Platform DevOps team keeps uptime of the cluster nodes high. This is neither practical at scale of CSPs nor it is conformant to cloud native practices.

Reviewer 3: I think we need to explore the concept of pinning. Pinning is a loaded term, and I'm pretty sure you are referring to a pod/CNF having its life-cycle tightly coupled to a specific node. Therefore we lose the ability to taint nodes and redeploy, recover from failures etc...

I think though that "pinning" resources in a node is not a bad thing if we can do it in a cloud native way. If the labels are there, and I can still meet the early considerations in this doc (here is a new SRIOV capable node, and FPGA, or any other feature etc...). If you can jump from the driver seat of one Ferrari to another, I think the intent is met.

Author: Correct, I meant here "node pinning" which is blocking the node from being tainted and drained. 

CPU pinning within the node is separate use case that I am not discussing here.

Author: I have sharpened it in the latest update. Please check if the additional wording makes it better.

Reviewer: @jeffsaelens it looks like this conversation is for the older version.  Please check to see if the wording (eg. pinning) and other items are addressed in the new version.  Marking this conversation as resolved.

> A CNF DevOps team shall be in position to quickly determine if CNF is behaving and functioning propperly after lifecycle operation (i.e. change) on infrastructure is done. If all is ok it is confirmation that it is ok to promote that change to next environment and ultimately to the production. With the normal lifecycle pace of Cloud Native Infrastructure CNF DevOps team needs to count that this ahs to be done 4-6 times a month for all ransitions between the environments. Ideally, it shall be happen automatically with combination of application testing and monitoring where the results can be available next day to CNF DevOps team. If that is not feasible it would still be acceptable if CNF DevOps team does that within 7 days per lifecycle operation.

Reviewer 5: ```suggestion
A CNF DevOps team shall be in position to quickly determine if CNF is behaving and functioning properly after lifecycle operation (i.e. change) on infrastructure is done. If all is ok it is confirmation that it is ok to promote that change to next environment and ultimately to the production. With the normal lifecycle pace of Cloud Native Infrastructure CNF DevOps team needs to count that this has to be done 4-6 times a month for all transitions between the environments. Ideally, it shall be happen automatically with combination of application testing and monitoring where the results can be available next day to CNF DevOps team. If that is not feasible it would still be acceptable if CNF DevOps team does that within 7 days per lifecycle operation.
```

> ## Challenges and limitations with Kube-native approach

Reviewer 2: Do we also define(d) Kube native on top of cloud native?

Reviewer: @CsatariGergely yes. that can go into the global glossary as its applicable across the board.

Reviewer: @fkautz would you like to help contributing the kubenative definition to the glossary?

Reviewer 4: I created #159 to track this

Also, most of CNFs today are following systems integrated approach. This means that integration of CNF version X with Platform version Y has been tested separately, possily as part of periodic "plug tests" or the certification procedures. It is typically CNF vendor to Platform vendor activity. Upstream is missed out in these certifications. Integrated systems approach and periodic integration testing with fixed versions unfortunately does not work in cloud native scenario where change is only constant. Versions of components are changing frequently and not upgrading means only accumulating the risks and technical debt which is very expensive from operations perspective. If changes are done less frequently they are typically bigger and bring higher probability of failures.

> 

Reviewer 3: I almost think a lot of this could be a use case in itself.

Author: hm, could be. If somebody makes dedicated use case we could offload this one here. However I would rather go merge this soon and change later when/if new case appears.

Author: more I think of that it looks to me like two distinct topics:
1. How to reliably and quickly determine that after infrastructure change a CNF that already runs on my platform still runs properly, so that I can promote infrastructure change to reference and ultimately to the production.
2. How do we certify that CNF can run on my platform and if the platform provides everything so that CNF vendor can offer certain SLA.

The later one will partially be one of the ultimate outcomes of CNF WG in a sense that if both CNF and infrastructure conform to set of best practices they will be able to run together.

Reviewer 5: Maybe it's important to distinguish the Platform requirements in this use case. Kubernetes API provides a well defined version management for its resources (alpha, beta, v1). But if the CNF has some non-Kubernetes requirements like Kernel version, CPU/memory features and/or PCI devices those can be addressed by the [NFD](https://kubernetes-sigs.github.io/node-feature-discovery/v0.8/get-started/introduction.html#feature-discovery). Maybe the best practice here is to encourage to use pinned versions.

Reviewer 2: [Anuket RA2](https://cntt.readthedocs.io/en/latest/ref_arch/kubernetes/) addresses the infrastructure related part of 2.
Kuberetes API defines only Kubernetes and tells nothing about different other capabilities of the cluster, so there is a need for more stuff to be defined.

Reviewer: @jeffsaelens what do you think about leaving the section on versions and system integration here and creating a new use case with that as a focus?  (eg. its okay for different uses to have common or duplicated content when it makes sense)

@vukg #2 this can be a focused use case where we talk about the different aspects of CNF and platform.  

Ideally a CNF is declares its needs and the platform will communicate what it can provide. The CNF then works with whats available. Of course there will be cases where a CNF may need specific features only available with K8s add-ons / extensions. The CNF declarations for these needs should exist, possibly with versions supported (ideally a version or any newer). The platform itself can be built following the same type of practices where the layers, add-ons/extensions, etc are all declared and versioned.

The long term goal is a negotiation of the application with the underlying platform to determine what is supported and adjust accordingly.

Reviewer 3: if we are ok with the duplications, then I'm down. At later dates, we could potentially make new PRs that link complimentary or parallel use cases.

Reviewer 4: do we want to create issues for the other use cases so people can be inspired by them then resolve this conversation and get this PR merged?

Author: @taylor today we are doing this alignment of platform (offer) and CNF (needs) in a dialog between platform and CNF teams+vendors as part of onboarding process. It includes itterations of platform configuration/extension and CNF addaptations and is mostly collaborative process that takes couple of days. However there could be also some more programatic / automated approach in future as "best practice".

> Also, most of CNFs today are following systems integrated approach. This means that integration of CNF version X with Platform version Y has been tested separately, possily as part of periodic "plug tests" or the certification procedures. It is typically CNF vendor to Platform vendor activity. Upstream is missed out in these certifications. Unfortunately, systems integrated approach and periodic integration testing with frozen versions, do not work in cloud native scenario where change is only constant. Versions of components are changing frequently and not upgrading means only accumulating the risks and technical debt. This is very expensive from operations perspective. If changes are done less frequently they are typically bigger and bring higher probability of failures.

Reviewer 5: ```suggestion
Also, most of CNFs today are following systems integrated approach. This means that integration of CNF version X with Platform version Y has been tested separately, possibly as part of periodic "plug tests" or the certification procedures. It is typically CNF vendor to Platform vendor activity. Upstream is missed out in these certifications. Unfortunately, systems integrated approach and periodic integration testing with frozen versions, do not work in cloud native scenario where change is only constant. Versions of components are changing frequently and not upgrading means only accumulating the risks and technical debt. This is very expensive from operations perspective. If changes are done less frequently they are typically bigger and bring higher probability of failures.
```

### Established practices to overcome challenges and limitations
N/A (tbd)

### What needs to be done differently in order to overcome challenges and limitations 
CNFs shall be written for cloud. This means that as boundary condition they shall assume that every node is ephemeral and short lived. They need to use native Kubernetes mechanisms to assure their own high availability in spite of dynamic nature of Kubernetes cluster in which they are running. They shall be able to recover from node failures and from node draining with marginal loss of traffic in the worst case and without visible impact on end customers or systems they serve.

> Each CNF shall come with suite of tests, testing scenarios and monitoring rules that CNF DevOps team can deploy in their continuous testing pipeline and monitoring system to enable that CNF is verified and certified (in best case daily) on that particular state of infrastructure.

Reviewer: @vukg suggested update for https://github.com/cncf/cnf-wg/pull/95#discussion_r603664833

```suggestion
Each CNF shall come with a suite of tests, testing scenarios and monitoring rules that CNF DevOps team can leverage. A team can deploy the test suite tooling via their continuous testing pipeline and integrate the results with their monitoring systems. The end goal being to ensure that a CNF is verified and certified (the best case being daily) on that particular state of infrastructure.
```

cc: @jeffsaelens @lixuna

Reviewer 6: LGTM

> Each CNF shall come with suite of tests, testing scenarios and monitoring rules that CNF DevOps team can deploy in their continous testing pipeline and monitoring system to enable that CNF is verified and certified (in best case daily) on that particular state of infrastructure.

Reviewer 5: ```suggestion
Each CNF shall come with suite of tests, testing scenarios and monitoring rules that CNF DevOps team can deploy in their continuous testing pipeline and monitoring system to enable that CNF is verified and certified (in best case daily) on that particular state of infrastructure.
```

A CNF DevOps team shall be in position to quickly determine if CNF is behaving and functioning propperly after lifecycle operation (i.e. change) on infrastructure is done. If all is ok it is confirmation that it is ok to promote that change to next environment and ultimately to the production. With the normal lifecycle pace of Cloud Native Infrastructure CNF DevOps team needs to count that this ahs to be done 4-6 times a month for all ransitions between the environments. Ideally, it shall be happen automatically with combination of application testing and monitoring where the results can be available next day to CNF DevOps team. If that is not feasible it would still be acceptable if CNF DevOps team does that within 7 days per lifecycle operation.

## Challenges and limitations with Kube-native approach

> Most of today's (so called) CNFs are retrofitted VNFs that have been forklifted into the containers. They are frequently pinning the nodes meaning that the CNF workloads are blocking that node from being drained in reasonably short timeframe. CNFs expect that the individual cluster nodes are long living, even highly available. In Kube-native approach cluster nodes are actually ephemeral and relatively short living. They exist typically for couple of days to couple of weeks. In such situation CNFs are implicitly requiring that infrastructure is mutable and that Cloud Native Platform DevOps team keeps uptime of the cluster nodes high. This is neither practical at scale of CSPs nor it is conformant to cloud native practices.

Reviewer: ```suggestion
Most of today's (so called) CNFs are retrofitted VNFs that have been forklifted into the containers.  The forklifted CNF is frequently tightly coupled to the underlying infrastructure, including going as far as assigning the CNF to a specific node. This means CNF workloads are blocking these nodes from being drained in reasonably short timeframe. These type of CNFs expect that the individual cluster nodes are long living, even highly available. In Kube-native approach cluster nodes are actually ephemeral and relatively short living. They exist typically for couple of days to couple of weeks. In such situation CNFs are implicitly requiring that infrastructure is mutable and that Cloud Native Platform DevOps team keeps uptime of the cluster nodes high. This is neither practical at scale of CSPs nor it is conformant to cloud native practices.
```

@jeffsaelens how's this for removing pinning?

@vukg thoughts?
